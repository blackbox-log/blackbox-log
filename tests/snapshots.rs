use std::collections::BTreeMap;
use std::fs;
use std::io::Read;

use blackbox_log::data::Stats;
use blackbox_log::frame::{self, FieldDef, GpsFrameDef};
use blackbox_log::prelude::*;
use blackbox_log::units::si;
use blackbox_log::{headers, Event, Unit, Value};
use hashbrown::HashMap;
use serde::ser::{SerializeStruct, Serializer};
use serde::Serialize;

const MAX_EVENTS: usize = 200_000;
const MAX_LOGS: usize = 2;

macro_rules! run {
    () => {
        |path| {
            let mut file = fs::File::open(path).unwrap();
            let mut data = Vec::new();
            file.read_to_end(&mut data).unwrap();

            let file = blackbox_log::File::new(&data);
            let snapshot = FileSnapshot::from(file);

            insta::assert_ron_snapshot!(snapshot);
        }
    };
}

#[test]
fn own() {
    insta::glob!("logs/*.bbl", run!());
}

#[test]
#[ignore]
fn fc_blackbox() {
    insta::glob!("logs/fc-blackbox/*", run!());
}

#[test]
#[ignore]
fn gimbal_ghost() {
    insta::glob!("logs/gimbal-ghost/*", run!());
}

#[derive(Debug, Serialize)]
#[serde(rename = "File")]
struct FileSnapshot<'data> {
    count: usize,
    logs: Vec<headers::ParseResult<LogSnapshot<'data>>>,
}

impl<'data> From<blackbox_log::File<'data>> for FileSnapshot<'data> {
    fn from(file: blackbox_log::File<'data>) -> Self {
        let logs = file
            .iter()
            .map(|mut reader| {
                Headers::parse(&mut reader).map(|headers| {
                    let data = DataParser::new(reader, &headers);
                    LogSnapshot::new(&headers, data)
                })
            })
            .take(MAX_LOGS)
            .collect::<Vec<_>>();

        Self {
            count: file.log_count(),
            logs,
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "Log")]
struct LogSnapshot<'data> {
    headers: HeadersSnapshot<'data>,
    stats: Stats,
    capped: bool,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    events: Vec<Event>,
    #[serde(skip_serializing_if = "MainSnapshot::is_empty")]
    main: MainSnapshot,
    #[serde(skip_serializing_if = "SlowSnapshot::is_empty")]
    slow: SlowSnapshot,
    #[serde(skip_serializing_if = "GpsSnapshot::is_empty")]
    gps: GpsSnapshot,
}

impl<'data> LogSnapshot<'data> {
    fn new(headers: &Headers<'data>, mut data: DataParser<'data, '_>) -> Self {
        let mut events = Vec::new();

        let main = headers.main_frame_def.iter().collect::<Fields>();
        let mut main = MainSnapshot::new(main);

        let slow = headers.slow_frame_def.iter().collect::<Fields>();
        let mut slow = SlowSnapshot::new(slow);

        let gps = headers
            .gps_frame_def
            .iter()
            .flat_map(GpsFrameDef::iter)
            .collect::<Fields>();
        let mut gps = GpsSnapshot::new(gps);

        let mut capped = true;
        for _ in 0..MAX_EVENTS {
            let Some(frame) = data.next() else {
                capped = false;
                break;
            };

            match frame {
                ParserEvent::Event(event) => events.push(event),
                ParserEvent::Main(frame) => main.update(frame),
                ParserEvent::Slow(frame) => slow.update(frame),
                ParserEvent::Gps(frame) => gps.update(frame),
            }
        }

        Self {
            headers: headers.into(),
            stats: data.stats().clone(),
            capped,
            events,
            main,
            slow,
            gps,
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "Headers")]
struct HeadersSnapshot<'data> {
    firmware_revision: &'data str,
    firmware: headers::Firmware,
    firmware_date: Option<Result<String, &'data str>>,
    board_info: Option<&'data str>,
    craft_name: Option<&'data str>,
    debug_mode: headers::DebugMode,
    disabled_fields: headers::DisabledFields,
    features: headers::FeatureSet,
    unknown: HashMap<&'data str, &'data str>,
}

impl<'data, 'a> From<&'a Headers<'data>> for HeadersSnapshot<'data> {
    fn from(headers: &'a Headers<'data>) -> Self {
        HeadersSnapshot {
            firmware_revision: headers.firmware_revision(),
            firmware: headers.firmware(),
            firmware_date: headers
                .firmware_date()
                .map(|res| res.map(|date| date.to_string())),
            board_info: headers.board_info(),
            craft_name: headers.craft_name(),
            debug_mode: headers.debug_mode(),
            disabled_fields: headers.disabled_fields(),
            features: headers.features(),
            unknown: headers.unknown().clone(),
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "MainFrame")]
struct MainSnapshot {
    count: u128,
    last_iteration: u32,
    time: NumberHistory<16, i128>,
    fields: Fields,
}

impl MainSnapshot {
    fn new(fields: Fields) -> Self {
        Self {
            count: 0,
            last_iteration: 0,
            time: NumberHistory::new(),
            fields,
        }
    }

    fn update(&mut self, frame: frame::MainFrame) {
        self.count += 1;
        self.last_iteration = frame.iteration();
        self.time.update(frame.time_raw().into());
        self.fields.update(frame);
    }

    fn is_empty(&self) -> bool {
        self.count == 0
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "SlowFrame")]
struct SlowSnapshot {
    count: u32,
    fields: Fields,
}

impl SlowSnapshot {
    fn new(fields: Fields) -> Self {
        Self { count: 0, fields }
    }

    fn update(&mut self, frame: frame::SlowFrame) {
        self.count += 1;
        self.fields.update(frame);
    }

    fn is_empty(&self) -> bool {
        self.count == 0
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "GpsFrame")]
struct GpsSnapshot {
    count: u32,
    time: (),
    fields: Fields,
}

impl GpsSnapshot {
    fn new(fields: Fields) -> Self {
        Self {
            count: 0,
            time: (),
            fields,
        }
    }

    fn update(&mut self, frame: frame::GpsFrame) {
        self.count += 1;
        self.fields.update(frame);
    }

    fn is_empty(&self) -> bool {
        self.count == 0
    }
}

#[derive(Debug, Serialize)]
struct Fields(Vec<FieldSnapshot>);

impl Fields {
    fn update<F: blackbox_log::frame::Frame>(&mut self, frame: F) {
        for (field, value) in self.0.iter_mut().zip(frame.iter()) {
            field.update(value.into());
        }
    }
}

impl<'data, U: Into<Unit>> FromIterator<FieldDef<'data, U>> for Fields {
    fn from_iter<I: IntoIterator<Item = FieldDef<'data, U>>>(iter: I) -> Self {
        Self(
            iter.into_iter()
                .map(|field| FieldSnapshot::new(field.name.to_owned(), field.unit.into()))
                .collect(),
        )
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "Field")]
struct FieldSnapshot {
    name: String,
    unit: Unit,
    history: History,
}

#[derive(Debug, Serialize)]
#[serde(untagged)]
enum History {
    Int(NumberHistory<16, i128>),
    Bool { yes: usize, no: usize },
    Flags(BTreeMap<&'static str, usize>),
}

#[derive(Debug)]
struct NumberHistory<const N: usize, T> {
    min: T,
    max: T,
    seen: Vec<T>,
    histogram: [usize; N],
}

impl FieldSnapshot {
    fn new(name: String, unit: Unit) -> Self {
        Self {
            name,
            unit,
            history: match unit {
                Unit::Amperage
                | Unit::Voltage
                | Unit::Acceleration
                | Unit::Rotation
                | Unit::GpsCoordinate
                | Unit::Altitude
                | Unit::Velocity
                | Unit::GpsHeading
                | Unit::Unitless => History::Int(NumberHistory::new()),
                Unit::Boolean => History::Bool { yes: 0, no: 0 },
                Unit::FlightMode | Unit::State | Unit::FailsafePhase => {
                    History::Flags(BTreeMap::new())
                }
            },
        }
    }

    #[allow(clippy::wildcard_enum_match_arm, clippy::cast_possible_truncation)]
    fn update(&mut self, value: Value) {
        match &mut self.history {
            History::Int(history) => history.update(match value {
                Value::Amperage(a) => a.get::<si::electric_current::milliampere>().round() as i128,
                Value::Voltage(v) => v.get::<si::electric_potential::millivolt>().round() as i128,
                Value::Acceleration(a) => a
                    .get::<si::acceleration::centimeter_per_second_squared>()
                    .round() as i128,
                Value::Rotation(r) => {
                    r.get::<si::angular_velocity::degree_per_second>().round() as i128
                }
                Value::GpsCoordinate(c) => (c * 10000000.).round() as i128,
                Value::Altitude(a) => a.get::<si::length::meter>().round() as i128,
                Value::Velocity(v) => {
                    v.get::<si::velocity::centimeter_per_second>().round() as i128
                }
                Value::GpsHeading(h) => (h * 10.).round() as i128,
                Value::Unsigned(u) => u.into(),
                Value::Signed(s) => s.into(),
                _ => unreachable!(),
            }),
            History::Bool { yes, no } => match value {
                Value::Boolean(true) => *yes += 1,
                Value::Boolean(false) => *no += 1,
                _ => unreachable!(),
            },
            History::Flags(history) => {
                if let Value::FailsafePhase(phase) = value {
                    *history.entry(phase.as_name()).or_insert(0) += 1;
                } else {
                    let flags = match value {
                        Value::FlightMode(m) => m.as_names(),
                        Value::State(s) => s.as_names(),
                        _ => unreachable!(),
                    };

                    for flag in flags {
                        *history.entry(flag).or_insert(0) += 1;
                    }
                }
            }
        }
    }
}

impl<const N: usize, T> NumberHistory<N, T> {
    fn new() -> Self
    where
        T: Default,
    {
        Self {
            min: T::default(),
            max: T::default(),
            seen: Vec::new(),
            histogram: [0; N],
        }
    }

    fn update_range(&mut self, value: T)
    where
        T: PartialOrd,
    {
        if value < self.min {
            self.min = value;
        } else if value > self.max {
            self.max = value;
        }
    }

    /// Store new value sorted in `seen` if not already present
    fn update_seen(&mut self, value: T)
    where
        T: PartialOrd,
    {
        let index = self.seen.partition_point(|x| *x < value);
        if self.seen.get(index) != Some(&value) {
            self.seen.insert(index, value);
        }
    }

    /// Bucket so that sequential values go in sequential buckets, since data is
    /// usually clustered
    #[inline(always)]
    fn update_histogram(&mut self, value: u128) {
        self.histogram[(value % N as u128) as usize] += 1;
    }
}

impl<const N: usize> NumberHistory<N, i128> {
    fn update(&mut self, value: i128) {
        self.update_range(value);
        self.update_seen(value);
        self.update_histogram(value.unsigned_abs());
    }
}

impl<const N: usize> Serialize for NumberHistory<N, i128>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(self, self.min, self.max, serializer)
    }
}

impl<const N: usize> Serialize for NumberHistory<N, f64>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(
            self,
            format!("{:.2}", self.min),
            format!("{:.2}", self.max),
            serializer,
        )
    }
}

#[inline(always)]
fn serialize<const N: usize, T, S, U>(
    snapshot: &NumberHistory<N, T>,
    min: U,
    max: U,
    serializer: S,
) -> Result<S::Ok, S::Error>
where
    S: Serializer,
    U: Serialize,
    [usize; N]: Serialize,
{
    let unique = snapshot.seen.len();

    let mut state = serializer.serialize_struct("FieldSnapshot", 4)?;

    state.serialize_field("min", &min)?;
    state.serialize_field("max", &max)?;
    state.serialize_field("unique", &unique)?;

    let histogram = "histogram";
    if unique > 1 {
        state.serialize_field(histogram, &snapshot.histogram)?;
    } else {
        state.skip_field(histogram)?;
    }

    state.end()
}
