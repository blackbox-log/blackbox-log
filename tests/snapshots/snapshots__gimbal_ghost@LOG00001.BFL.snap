---
source: tests/snapshots.rs
expression: snapshot
input_file: ".modules/gimbal-ghost/test/LOG00001.BFL"
---
FileSnapshot(
  count: 1,
  logs: [
    Ok(LogSnapshot(
      headers: Headers(
        firmware_revision: "Betaflight 4.3.0 (fc8318f54) STM32F405",
        firmware: Betaflight("4.3.0"),
        board_info: Some("MTKS MATEKF405CTR"),
        craft_name: None,
        debug_mode: GyroScaled,
        disabled_fields: [],
        features: [
          RxSerial,
          SoftSerial,
          Gps,
          Telemetry,
          Osd,
          AirMode,
          AntiGravity,
        ],
        vbat_reference: Some(1629),
        acceleration_1g: Some(2048),
        gyro_scale: Some(0.017453292),
        min_throttle: Some(1070),
        motor_output_range: Some(MotorOutputRange(
          min: 158,
          max: 2047,
        )),
        unknown: {
          "acc_hardware": "0",
          "dterm_notch_cutoff": "0",
          "pid_process_denom": "1",
          "gyro_cal_on_first_arm": "0",
          "yaw_lowpass_hz": "100",
          "acc_limit_yaw": "0",
          "dyn_notch_max_hz": "600",
          "yaw_deadband": "0",
          "mag_hardware": "1",
          "simplified_dterm_filter_multiplier": "100",
          "dshot_bidir": "1",
          "dterm_lpf1_dyn_hz": "75,150",
          "simplified_feedforward_gain": "100",
          "gyro_sync_denom": "1",
          "simplified_dterm_filter": "1",
          "serialrx_provider": "12",
          "throttle_limit_type": "0",
          "dshot_idle_value": "550",
          "dyn_idle_p_gain": "50",
          "rc_smoothing_debug_axis": "0",
          "motor_output_limit": "100",
          "gyro_to_use": "0",
          "dterm_lpf2_type": "0",
          "vbatcellvoltage": "330,350,430",
          "tpa_breakpoint": "1350",
          "rpm_filter_harmonics": "3",
          "throttle_limit_percent": "100",
          "rates": "67,67,67",
          "acc_limit": "0",
          "d_min": "30,34,0",
          "feedforward_averaging": "0",
          "gyro_lpf1_static_hz": "250",
          "d_max_gain": "37",
          "dyn_idle_min_rpm": "0",
          "gyro_lpf1_dyn_hz": "250,500",
          "motor_poles": "14",
          "acc_lpf_hz": "1000",
          "vbat_sag_compensation": "0",
          "baro_hardware": "0",
          "rc_rates": "7,7,7",
          "pid_at_min_throttle": "1",
          "d_max_advance": "20",
          "P interval": "8",
          "rates_type": "3",
          "dyn_idle_i_gain": "50",
          "pidsum_limit_yaw": "400",
          "pidsum_limit": "500",
          "dyn_notch_min_hz": "150",
          "motor_pwm_rate": "480",
          "rc_smoothing_auto_factor_throttle": "30",
          "dyn_notch_q": "500",
          "dterm_notch_hz": "0",
          "rollPID": "45,80,40",
          "rc_smoothing_setpoint_cutoff": "0",
          "yawPID": "45,80,0",
          "levelPID": "50,50,75",
          "airmode_activate_throttle": "25",
          "rc_smoothing_feedforward_cutoff": "0",
          "gyro_hardware_lpf": "0",
          "anti_gravity_gain": "3500",
          "rc_smoothing_rx_average": "9059",
          "dyn_idle_max_increase": "150",
          "feedforward_max_rate_limit": "90",
          "rc_smoothing_auto_factor": "30",
          "iterm_windup": "85",
          "gyro_notch_cutoff": "0,0",
          "throttle_boost_cutoff": "15",
          "simplified_pitch_d_gain": "100",
          "rc_smoothing_active_cutoffs_ff_sp_thr": "41,41,41",
          "deadband": "0",
          "tpa_rate": "65",
          "rate_limits": "1998,1998,1998",
          "simplified_d_gain": "100",
          "rpm_filter_fade_range_hz": "50",
          "maxthrottle": "2000",
          "feedforward_smooth_factor": "25",
          "I interval": "256",
          "use_integrated_yaw": "0",
          "simplified_dmax_gain": "100",
          "simplified_gyro_filter": "1",
          "gyro_lpf1_type": "0",
          "rpm_filter_min_hz": "100",
          "dyn_idle_d_gain": "50",
          "simplified_pids_mode": "2",
          "Log start datetime": "2022-02-01T16:16:32.235+00:00",
          "gyro_lpf2_type": "0",
          "anti_gravity_mode": "0",
          "simplified_master_multiplier": "100",
          "iterm_relax": "1",
          "dyn_notch_count": "1",
          "gyro_notch_hz": "0,0",
          "iterm_relax_cutoff": "15",
          "simplified_pitch_pi_gain": "100",
          "rc_smoothing_throttle_cutoff": "0",
          "feedforward_boost": "15",
          "dterm_lpf1_type": "0",
          "motor_pwm_protocol": "7",
          "gyro_lpf2_static_hz": "500",
          "anti_gravity_threshold": "250",
          "rpm_filter_q": "500",
          "rc_smoothing": "1",
          "dterm_lpf1_static_hz": "75",
          "pitchPID": "47,84,46",
          "simplified_pi_gain": "100",
          "magPID": "40",
          "simplified_gyro_filter_multiplier": "100",
          "feedforward_jitter_factor": "7",
          "ff_weight": "120,125,120",
          "Firmware date": "Jan 30 2022 01:43:14",
          "iterm_relax_type": "1",
          "simplified_i_gain": "100",
          "rpm_filter_lpf_hz": "150",
          "thr_mid": "50",
          "throttle_boost": "5",
          "feedforward_transition": "0",
          "thr_expo": "0",
          "looptime": "125",
          "rc_expo": "0,0,0",
          "P ratio": "32",
          "dterm_lpf2_static_hz": "150",
          "use_unsynced_pwm": "0",
          "abs_control_gain": "0",
        },
      ),
      stats: Stats(
        counts: FrameCounts(
          event: 1,
          main: 198955,
          slow: 29,
          gps: 1015,
          gps_home: 0,
        ),
      ),
      capped: true,
      events: [
        SyncBeep(385573225),
      ],
      main: MainSnapshot(
        count: 198955,
        last_iteration: 1591562,
        time: FieldSnapshot(
          min: 0,
          max: 589915901,
          unique: 198955,
          histogram: (12296, 12414, 12374, 12409, 12440, 12433, 12334, 12471, 12361, 12483, 12257, 12552, 12638, 12457, 12434, 12602),
        ),
        fields: Fields([
          FieldSnapshot(
            name: "axisP[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -55,
              max: 56,
              unique: 100,
              histogram: (40753, 36244, 28945, 22454, 16886, 13003, 10014, 7785, 5965, 4527, 3594, 2773, 2157, 1643, 1213, 999),
            ),
          ),
          FieldSnapshot(
            name: "axisP[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -44,
              max: 44,
              unique: 86,
              histogram: (42995, 36846, 29096, 22303, 16983, 12591, 9612, 7428, 5544, 4221, 3200, 2472, 1940, 1583, 1198, 943),
            ),
          ),
          FieldSnapshot(
            name: "axisP[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -32,
              max: 21,
              unique: 51,
              histogram: (94508, 53546, 25185, 11568, 5779, 3281, 1706, 1001, 659, 546, 400, 287, 172, 135, 115, 67),
            ),
          ),
          FieldSnapshot(
            name: "axisI[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -58,
              max: 69,
              unique: 128,
              histogram: (23362, 19218, 18211, 18421, 18048, 16489, 16227, 13122, 10876, 8458, 7336, 7137, 6300, 6075, 4960, 4715),
            ),
          ),
          FieldSnapshot(
            name: "axisI[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -67,
              max: 111,
              unique: 179,
              histogram: (16471, 13337, 13370, 12536, 11535, 10679, 12210, 13604, 13822, 13080, 13627, 13423, 12292, 10445, 9981, 8543),
            ),
          ),
          FieldSnapshot(
            name: "axisI[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -84,
              max: 112,
              unique: 197,
              histogram: (20496, 15795, 14909, 14629, 15545, 15516, 13684, 13137, 11683, 11833, 9361, 9597, 9456, 8763, 7239, 7312),
            ),
          ),
          FieldSnapshot(
            name: "axisD[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -115,
              max: 140,
              unique: 158,
              histogram: (27258, 25725, 23569, 20857, 17919, 15209, 12754, 10977, 9425, 7729, 6529, 5718, 4918, 4006, 3438, 2924),
            ),
          ),
          FieldSnapshot(
            name: "axisD[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -76,
              max: 74,
              unique: 120,
              histogram: (34014, 31438, 26937, 22439, 17993, 14381, 11449, 9202, 7408, 5883, 4796, 3846, 3031, 2493, 2021, 1624),
            ),
          ),
          FieldSnapshot(
            name: "axisF[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -30,
              max: 38,
              unique: 69,
              histogram: (166686, 12460, 7843, 4555, 2595, 1592, 1061, 618, 467, 325, 230, 176, 153, 74, 64, 56),
            ),
          ),
          FieldSnapshot(
            name: "axisF[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -17,
              max: 20,
              unique: 38,
              histogram: (177320, 9317, 6106, 2581, 1504, 928, 455, 270, 120, 115, 55, 38, 34, 46, 33, 33),
            ),
          ),
          FieldSnapshot(
            name: "axisF[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -49,
              max: 33,
              unique: 71,
              histogram: (180325, 7169, 4621, 2666, 1313, 968, 606, 319, 227, 186, 142, 138, 139, 72, 41, 23),
            ),
          ),
          FieldSnapshot(
            name: "rcCommand[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -300,
              max: 171,
              unique: 331,
              histogram: (26441, 12954, 22032, 8792, 14456, 7931, 13396, 6352, 16741, 6411, 16099, 8299, 13489, 6089, 13413, 6060),
            ),
          ),
          FieldSnapshot(
            name: "rcCommand[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -159,
              max: 120,
              unique: 233,
              histogram: (40294, 15349, 22340, 8453, 17511, 7223, 13837, 8340, 13034, 3821, 11874, 4095, 11868, 4458, 12643, 3815),
            ),
          ),
          FieldSnapshot(
            name: "rcCommand[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -119,
              max: 249,
              unique: 288,
              histogram: (22075, 19452, 24631, 7241, 20665, 7189, 13467, 6567, 14878, 7358, 13861, 5084, 14118, 5626, 10435, 6308),
            ),
          ),
          FieldSnapshot(
            name: "rcCommand[3]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 1844,
              unique: 837,
              histogram: (10168, 13156, 19285, 10592, 13292, 12571, 8569, 13780, 15102, 11635, 18045, 9823, 9132, 14599, 12172, 7034),
            ),
          ),
          FieldSnapshot(
            name: "setpoint[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -258,
              max: 94,
              unique: 332,
              histogram: (30317, 14313, 14436, 16834, 13349, 10551, 11321, 9293, 8612, 10593, 13728, 10416, 9620, 6924, 8635, 10013),
            ),
          ),
          FieldSnapshot(
            name: "setpoint[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -83,
              max: 51,
              unique: 135,
              histogram: (47937, 21427, 13960, 17926, 13474, 10114, 8684, 8375, 7123, 8313, 6338, 8352, 8526, 6222, 6099, 6085),
            ),
          ),
          FieldSnapshot(
            name: "setpoint[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -50,
              max: 183,
              unique: 230,
              histogram: (38301, 21277, 16133, 11297, 11265, 9210, 9458, 12475, 11265, 7502, 8797, 8498, 10582, 8300, 7585, 7010),
            ),
          ),
          FieldSnapshot(
            name: "setpoint[3]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 848,
              unique: 837,
              histogram: (14920, 11458, 18303, 10256, 9168, 14327, 11911, 7252, 10150, 13450, 18858, 10485, 13132, 12632, 8620, 14033),
            ),
          ),
          FieldSnapshot(
            name: "vbatLatest",
            unit: Voltage,
            history: FieldSnapshot(
              min: 0,
              max: 17520,
              unique: 433,
              histogram: (23578, 0, 24171, 0, 25741, 0, 24735, 0, 25478, 0, 24049, 0, 26603, 0, 24600, 0),
            ),
          ),
          FieldSnapshot(
            name: "amperageLatest",
            unit: Amperage,
            history: FieldSnapshot(
              min: 0,
              max: 75580,
              unique: 764,
              histogram: (25315, 0, 25499, 0, 27354, 0, 23536, 0, 24970, 0, 24490, 0, 23634, 0, 24157, 0),
            ),
          ),
          FieldSnapshot(
            name: "BaroAlt",
            unit: Unitless,
            history: FieldSnapshot(
              min: -381,
              max: 9556,
              unique: 4899,
              histogram: (12810, 13315, 12101, 12379, 12230, 12491, 12079, 12372, 11802, 13608, 12982, 11913, 11990, 13414, 11480, 11989),
            ),
          ),
          FieldSnapshot(
            name: "rssi",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 918,
              unique: 521,
              histogram: (14382, 10845, 14721, 12161, 11935, 13967, 10274, 15580, 9818, 14644, 11007, 12004, 12706, 10358, 14225, 10328),
            ),
          ),
          FieldSnapshot(
            name: "gyroADC[0]",
            unit: Rotation,
            history: FieldSnapshot(
              min: -270,
              max: 97,
              unique: 354,
              histogram: (15863, 22164, 18307, 16215, 14481, 13065, 11509, 10799, 10292, 10441, 10428, 10425, 9554, 8824, 8426, 8162),
            ),
          ),
          FieldSnapshot(
            name: "gyroADC[1]",
            unit: Rotation,
            history: FieldSnapshot(
              min: -103,
              max: 64,
              unique: 166,
              histogram: (21012, 31294, 24136, 18720, 15044, 12581, 10688, 9848, 9023, 8507, 8092, 7162, 6633, 5914, 5389, 4912),
            ),
          ),
          FieldSnapshot(
            name: "gyroADC[2]",
            unit: Rotation,
            history: FieldSnapshot(
              min: -55,
              max: 190,
              unique: 244,
              histogram: (23321, 32069, 20958, 14225, 10744, 9707, 9312, 9942, 9632, 9176, 9360, 9614, 9551, 8410, 6527, 6407),
            ),
          ),
          FieldSnapshot(
            name: "accSmooth[0]",
            unit: Acceleration,
            history: FieldSnapshot(
              min: -1554,
              max: 445,
              unique: 1833,
              histogram: (12614, 12860, 13144, 12625, 12774, 12293, 12341, 12943, 12222, 11985, 11980, 12389, 12121, 12286, 12194, 12184),
            ),
          ),
          FieldSnapshot(
            name: "accSmooth[1]",
            unit: Acceleration,
            history: FieldSnapshot(
              min: -1045,
              max: 1362,
              unique: 2158,
              histogram: (12393, 12480, 12913, 12434, 12641, 12458, 12669, 13019, 12619, 12613, 12083, 12659, 12193, 12086, 11981, 11714),
            ),
          ),
          FieldSnapshot(
            name: "accSmooth[2]",
            unit: Acceleration,
            history: FieldSnapshot(
              min: -298,
              max: 4576,
              unique: 4298,
              histogram: (12369, 12412, 12537, 12432, 12395, 12438, 12393, 12244, 12347, 12591, 12660, 12456, 12303, 12501, 12427, 12450),
            ),
          ),
          FieldSnapshot(
            name: "debug[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -471,
              max: 332,
              unique: 637,
              histogram: (13236, 15541, 14731, 13644, 13471, 13301, 12775, 12330, 12154, 12026, 11798, 11314, 10976, 10914, 10470, 10274),
            ),
          ),
          FieldSnapshot(
            name: "debug[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -155,
              max: 103,
              unique: 235,
              histogram: (13386, 18023, 16482, 15285, 14920, 14304, 13275, 12737, 12139, 11639, 10880, 10431, 9759, 8994, 8634, 8067),
            ),
          ),
          FieldSnapshot(
            name: "debug[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -93,
              max: 225,
              unique: 301,
              histogram: (13352, 17670, 15845, 14868, 14565, 13841, 13244, 12552, 11910, 11519, 11041, 10387, 10130, 9765, 9332, 8934),
            ),
          ),
          FieldSnapshot(
            name: "debug[3]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          FieldSnapshot(
            name: "motor[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 2057,
              unique: 1873,
              histogram: (12269, 12536, 12411, 12449, 12433, 12082, 12443, 12484, 12378, 12322, 12296, 12378, 12508, 12599, 12876, 12491),
            ),
          ),
          FieldSnapshot(
            name: "motor[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 1848,
              unique: 1579,
              histogram: (12501, 12322, 12274, 12250, 12618, 12454, 12660, 12110, 12327, 12356, 12357, 12619, 12609, 12493, 12588, 12417),
            ),
          ),
          FieldSnapshot(
            name: "motor[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 1986,
              unique: 1769,
              histogram: (12309, 12157, 12368, 12370, 12381, 12536, 12319, 12317, 12372, 12417, 12459, 12289, 12627, 12806, 12603, 12625),
            ),
          ),
          FieldSnapshot(
            name: "motor[3]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 1851,
              unique: 1656,
              histogram: (12510, 12408, 12425, 12464, 12374, 12371, 12244, 12355, 12514, 12361, 12534, 12638, 12369, 12427, 12536, 12425),
            ),
          ),
        ]),
      ),
      slow: SlowSnapshot(
        count: 29,
        fields: Fields([
          FieldSnapshot(
            name: "flightModeFlags",
            unit: FlightMode,
            history: {
              "ARM": 29,
            },
          ),
          FieldSnapshot(
            name: "stateFlags",
            unit: State,
            history: {
              "GPS_FIX": 26,
              "GPS_FIX_EVER": 29,
            },
          ),
          FieldSnapshot(
            name: "failsafePhase",
            unit: FailsafePhase,
            history: {
              "IDLE": 29,
            },
          ),
          FieldSnapshot(
            name: "rxSignalReceived",
            unit: Boolean,
            history: History(
              yes: 29,
              no: 0,
            ),
          ),
          FieldSnapshot(
            name: "rxFlightChannelsValid",
            unit: Boolean,
            history: History(
              yes: 29,
              no: 0,
            ),
          ),
        ]),
      ),
      gps: GpsSnapshot(
        count: 1015,
        time: (),
        fields: Fields([
          FieldSnapshot(
            name: "GPS_numSat",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 6,
              unique: 4,
              histogram: (37, 0, 0, 0, 481, 250, 247, 0, 0, 0, 0, 0, 0, 0, 0, 0),
            ),
          ),
          FieldSnapshot(
            name: "GPS_coord[0]",
            unit: GpsCoordinate,
            history: FieldSnapshot(
              min: -337981507,
              max: 0,
              unique: 953,
              histogram: (63, 60, 58, 72, 66, 83, 62, 68, 60, 68, 64, 47, 63, 57, 64, 60),
            ),
          ),
          FieldSnapshot(
            name: "GPS_coord[1]",
            unit: GpsCoordinate,
            history: FieldSnapshot(
              min: 0,
              max: 201019858,
              unique: 966,
              histogram: (68, 56, 74, 61, 61, 56, 62, 78, 64, 73, 60, 52, 51, 68, 64, 67),
            ),
          ),
          FieldSnapshot(
            name: "GPS_altitude",
            unit: Altitude,
            history: FieldSnapshot(
              min: 0,
              max: 3747,
              unique: 669,
              histogram: (64, 58, 62, 68, 47, 57, 61, 64, 75, 55, 54, 72, 62, 53, 87, 76),
            ),
          ),
          FieldSnapshot(
            name: "GPS_speed",
            unit: Velocity,
            history: FieldSnapshot(
              min: 0,
              max: 2806,
              unique: 802,
              histogram: (77, 82, 67, 51, 59, 60, 74, 67, 68, 62, 64, 68, 57, 54, 58, 47),
            ),
          ),
          FieldSnapshot(
            name: "GPS_ground_course",
            unit: GpsHeading,
            history: FieldSnapshot(
              min: 0,
              max: 3598,
              unique: 740,
              histogram: (64, 57, 99, 76, 88, 43, 79, 52, 62, 58, 75, 38, 57, 51, 52, 64),
            ),
          ),
        ]),
      ),
    )),
  ],
)
